src/Ads.py

```diff
@@ -14,87 +18,153 @@ class Reimbursement(object):
+        results['adjusted_cost'] = results['cost'].astype(float) * results['cost_share_rate']
 
-    def config(self, config_dict):
-        '''
-        config_dict has:
-            ad_type:
-            cost_share_rate:
-            allowed_spend_per_ad: (range) - assumption made: this is what is allowed to be reimbursed
-        '''
+        # compute the reimbursement.  Leave zero reimbursement until last to simplify
+        reimb_idx = results['adjusted_cost'] >= results['minimum_spend_per_ad']
+        results.loc[reimb_idx, 'reimbursement'] = \
+            results.loc[reimb_idx].apply(lambda rec: min(rec['adjusted_cost'], rec['maximum_spend_per_ad'] - 1), axis='columns')
+        # zero reimbursement as it didn't achieve the minimum ad spend
+        no_reimb_idx = results['adjusted_cost'] < results['minimum_spend_per_ad']
+        results.loc[no_reimb_idx, 'reimbursement'] = 0.0
 
-        self.configuration = config_dict
+        return results
 
-    def compute(self):
+    def set_ads(self, ads):
+        self.ads = ads
+
+    def get_ads(self):
+        return self.ads
+
+    def get_ad_types(self):
+        return set([ad['type'] for ad in self.ads])
+
+    def filter_ad_type(self, filter_set):
         '''
-        compute the statistics on the cost share rate and reimbursement amounts per ad_type
+        filter_set: set, from a subset of valid ad_types
+
+        returns new Reimbursement object that has ads from current Reimbursement object
+        where the ad_type of the included Ads matches any of the ad types in filter
 
-        Ad(0011) - cost share rate = .50 and spend per ad is $200(reimbursement amount)
-        Ad(0011)
         '''
+        filter_set = filter_set.intersection(set(self.parameters['ad_type']))
+        matched_ads = [ad for ad in self.ads if ad['type'] in filter_set]
+        return Reimbursement(self.parameters, matched_ads)
 
     def __str__(self):
         return '\n'.join([str(ad) for ad in self.ads])
-    
 
 
 class Ads(list):
     '''
     This is a list of Ad objects
     '''
 
-    def __init__(self):
-        super().__init__(self)
+    dataframe = None
 
+    def __init__(self, ad_list=[]):
+        '''
+        ad_list: list of Ad objects
+        '''
+        super().__init__(ad_list)
+        self.set_ad_types()
 
-class Ad(object):
+    def set_ad_types(self):
+        if len(self) > 0:
+            self.types = pd.DataFrame(self)['type'].value_counts().to_dict()
+        else:
+            self.types = None
 
-    VALID_AD_TYPES = set(['0011', '1011', '1111', '1010'])
+    def __str__(self):
+        return '\n'.join([str(ad) for ad in self])
 
-    def __init__(self, ad_type, ad_name):
+    def get_ad_types(self):
+        self.set_ad_types()
+        return self.types
 
-        self.ad_type = ad_type
-        self.ad_name = ad_name
 
-    def __str__(self):
-        return f'Ad_type: {self.ad_type} ... {self.ad_name}'
+class Ad(dict):
```

Comment:
Why use "dict" as the base class?

One-to-many Ad objects are used to compose an Ads object.
Since Ads will be analyzed using a dataframe, it is very easy to then create that dataframe by just processing a list of dicts.  e.g. pd.DataFrame(ads)

A dataframe is easy to use to analyze, filter, apply formulae (like reimbursement computation)
 
One alternative would have been to just use a standard "object" as the base-class and represent construct the Ads dataframe a bit differently.  But this seemed more elegant to me.